Blockbase: Add social navigation to blockbase themes

scruffian opened the pull request:
#### Changes proposed in this Pull Request:
This adds social navigation to all Blockbase themes:

Seedlet blocks
<img width="569" alt="Screenshot 2021-10-12 at 14 09 19" src="https://user-images.githubusercontent.com/275961/136962477-1cbdde2f-4c01-4abb-92de-8ec3690c3a1c.png">
<img width="475" alt="Screenshot 2021-10-12 at 14 09 08" src="https://user-images.githubusercontent.com/275961/136962487-c50ddea5-86d5-445c-9fcb-91385d5c2fe7.png">

Mayland Blocks
<img width="1109" alt="Screenshot 2021-10-12 at 14 08 33" src="https://user-images.githubusercontent.com/275961/136962495-75314e4e-0b32-42c6-8104-01340111a6de.png">
<img width="477" alt="Screenshot 2021-10-12 at 14 08 45" src="https://user-images.githubusercontent.com/275961/136962493-4e9b3dc9-6e2f-4d16-ad60-edcbc7b711b7.png">

Geologist
<img width="1073" alt="Screenshot 2021-10-12 at 14 07 30" src="https://user-images.githubusercontent.com/275961/136962500-d5e0f37c-0c2c-41aa-95ba-b7ccf7e555cb.png">
<img width="440" alt="Screenshot 2021-10-12 at 14 07 45" src="https://user-images.githubusercontent.com/275961/136962498-952ec1ac-894e-4ff1-9847-aa614ced280a.png">

Quadrat
<img width="1114" alt="Screenshot 2021-10-12 at 14 05 54" src="https://user-images.githubusercontent.com/275961/136962503-7021f499-eea8-4529-9958-1773636acbf0.png">
<img width="508" alt="Screenshot 2021-10-12 at 14 06 08" src="https://user-images.githubusercontent.com/275961/136962502-92ebae6f-9ad8-40c4-81f1-fc76a5d4fbbc.png">

Blockbase:
<img width="1092" alt="Screenshot 2021-10-12 at 14 03 06" src="https://user-images.githubusercontent.com/275961/136962509-46b78a56-8401-4c67-ba99-c641b0960de5.png">
<img width="500" alt="Screenshot 2021-10-12 at 14 03 20" src="https://user-images.githubusercontent.com/275961/136962505-7c6aab86-a293-4a27-8dda-ebf21ad46bb0.png">

Skatepark
<img width="878" alt="Screenshot 2021-10-12 at 13 25 04" src="https://user-images.githubusercontent.com/275961/136962515-9ce9c776-0dbd-4c26-a221-09012d17cd9c.png">
<img width="439" alt="Screenshot 2021-10-12 at 13 25 39" src="https://user-images.githubusercontent.com/275961/136962512-fc39072c-9881-4f83-a7f5-0b2156dc84e3.png">
<img width="439" alt="Screenshot 2021-10-12 at 13 25 14" src="https://user-images.githubusercontent.com/275961/136962513-d82b7a20-66c1-478e-af66-518b79866e4e.png">

scruffian committed:
Blockbase: Add social navigation to blockbase themes

jffng commented:
I rebased this now that #4468 is merged. Seedlet and Mayland look okay, but I have two questions about Blockbase and Quadrat:

1. How should we handle alignment? 

<img width="1440" alt="Screenshot 2021-08-27 at 13 13 13" src="https://user-images.githubusercontent.com/275961/131125957-2f648a78-9980-4976-b886-e24244de8c36.png">

One option is adding a `margin-left: auto;` to the primary nav in CSS. 

2. How should we handle the header when the responsive menu option is enabled? 

<img width="335" alt="Screen Shot 2021-09-01 at 3 52 13 PM" src="https://user-images.githubusercontent.com/5375500/131735314-39f123fa-e30d-4354-bc1f-f9eec22cd79c.png">
<img width="333" alt="Screen Shot 2021-09-01 at 3 54 13 PM" src="https://user-images.githubusercontent.com/5375500/131735380-d1601c8d-0bbf-456b-9e90-673e534daa2b.png">

scruffian commented:
@kjellr, do you have any thoughts on this?

kjellr commented:
> How should we handle alignment?
> One option is adding a margin-left: auto; to the primary nav in CSS.

For desktop, I think we should do that. In general, this should look like the Quadrat screenshot above, but with more margin in between the nav + the social links:

<img width="235" alt="Screen Shot 2021-09-02 at 12 13 18 PM" src="https://user-images.githubusercontent.com/1202812/131879821-2614bd4d-91a6-4ba9-a57f-1237a5aecf3e.png">
 
The right margin should be equal to the gap between nav items. 

> How should we handle the header when the responsive menu option is enabled?

This is another example of where we would ideally put these in the expanded hamburger menu. Part of me wonders if we should hack some temporary jQuery in there to move it there on small screens. 😅 Let me think about this on a little and get back to you later today/tomorrow.

kjellr commented on seedlet-blocks/block-template-parts/header.html:
```diff
@@ -11,6 +11,8 @@
 <!-- wp:navigation {"itemsJustification":"center","isResponsive":true,"__unstableLocation":"primary"} -->
 <!-- /wp:navigation -->
 
+<!-- wp:navigation {"orientation":"horizontal","itemsJustification":"center","isResponsive":true,"__unstableLocation":"social"} --><!-- /wp:navigation -->
```
For Seedlet, let's use a default icon color of black (the foreground color) to match the social menu color in the classic version of the theme.

kjellr commented:
@melchoyce @beafialho what do you think of these options for social links placement on small screens in Blockbase and Quadrat? 

![Frame 2](https://user-images.githubusercontent.com/1202812/132053857-431330dc-2dfd-4a65-b803-93b90f75959b.png)

None of them seem ideal to me, so I've also offered [a plea for a better solution in the issue here](https://href.li/?https://github.com/WordPress/gutenberg/issues/33278#issuecomment-912743391).

beafialho commented:
I think the best and less invasive option is definitely the top middle one, it makes most sense for the social icons to be beneath the hamburger. The lines add way too much noise to the top area (in this case, where these themes don't use them anywhere else).

I agree that the solution you propose in [the Navigation block issue](https://github.com/WordPress/gutenberg/issues/33278#issuecomment-876411953) should be possible as a default.

kjellr commented:
Ok, looping back to this. It _is_ possible to put social links within a navigation block, so we should just aim to do that instead of having two separate navigation menus.

<img width="601" alt="Screen Shot 2021-09-07 at 8 08 12 AM" src="https://user-images.githubusercontent.com/1202812/132342139-f970d467-8206-4f9b-95ea-b44b9e041dbe.png">

![mobile-nav](https://user-images.githubusercontent.com/1202812/132342093-827c8f2d-f506-4bdc-8571-87fb8f522b2d.gif)

scruffian committed:
Add social navigation links to primary navigation

scruffian commented:
I updated the approach here - now the social navigation gets appended to the primary navigation block. This means that it gets wrapped by the hamburger menu when in smaller screens.

We probably need some CSS to fix the alignments of the social menu, but I'm curious to get feedback on the approach before we do that.

pbking commented:
I **really** like the approach.  

I kind of would prefer to have a secondary attribute to key on for appending the social menu to the block.
(i.e. `<-- wp:navigation {__unstableLocation:"primary",__unstableAppendSocial=true} -->` or the ability to set __unstableLocation multiple times, or to an array or something)

But presently I think maybe that's something the feature potentially could grow into rather than start at.

It feels like the cleanest way to achieve the design target.

Only beef I have with the code is a preference to move the menu rendering stuff to another file.

scruffian committed:
move to a different file

scruffian commented:
> Only beef I have with the code is a preference to move the menu rendering stuff to another file.

Good idea, done!

scruffian committed:
add missing file

MaggieCabrera commented:
I'm getting this error when I apply this branch and go to the customizer and add a social menu to the "Social" location:

> ( ! ) Warning: DOMDocument::loadHTML(): Tag nav invalid in Entity, line: 1 in /Users/maggie/A8c/repos/a8c-projects/themes/blockbase/inc/social-navigation.php on line 48

MaggieCabrera commented on blockbase/inc/social-navigation.php:
```diff
@@ -0,0 +1,70 @@
+
+// We should only change the render of the navigtion block
+// to social links in the following conditions.
+function blockbase_condition_to_render_social_menu( $block ) {
```
A theme may want to define a menu that follows these conditions to show up in the footer too for example, or a pre-footer template part. In some cases we may want to have the social navigation but maybe not always. The perfect example of this is actually Skatepark, with this PR in place, this is what I see when I insert the pre footer block pattern:

<img width="1310" alt="Screenshot 2021-10-04 at 10 30 22" src="https://user-images.githubusercontent.com/3593343/135818923-fb933887-6861-423f-a0d9-87a87cc04a38.png">

scruffian committed:
suppress warnings for HTML5 elements

scruffian committed:
only append social links to a navigation block if the block opts in

scruffian commented:
> I'm getting this error when I apply this branch and go to the customizer and add a social menu to the "Social" location:

Fixed in 21c1ed4

scruffian commented:
> I kind of would prefer to have a secondary attribute to key on for appending the social menu to the block.
(i.e. <-- wp:navigation {__unstableLocation:"primary",__unstableAppendSocial=true} --> or the ability to set __unstableLocation multiple times, or to an array or something)

This turned out to be a very good idea, as it means we can control which menus the links get appended to, as I have done in 3a9f0f5.

If y'all like this approach we can add it to the header nav in all blockbase themes.

MaggieCabrera commented:
Now it's fataling for me:
<img width="1292" alt="Screenshot 2021-10-05 at 18 05 27" src="https://user-images.githubusercontent.com/3593343/136060609-87ff05d7-1e9c-41be-b929-529b9f55d7b8.png">

scruffian committed:
Fix fatal

scruffian commented:
I fixed the fatal, so you should be able to test now.

This revealed one shortcoming of this approach - if the primary navigation location is empty then no navigation is output, even if something is set in social...

MaggieCabrera commented:
> I fixed the fatal, so you should be able to test now.
> 
> This revealed one shortcoming of this approach - if the primary navigation location is empty then no navigation is output, even if something is set in social...

Yes! no more errors for me. Now that we have `__unstableSocialLinks`, why do we need to check for the other attribute? I think having `__unstableSocialLinks` is specific enough for us to know that we want to render the social links

RavanH commented:
Hi all, I've been testing the current social menu implementation in trunk with a local child theme and it seems to work well. Only concern I have is that (unlike item justification) it's not possible to change/unset icon color and icon size.

Is this feature considered already, and if so, is anybody working on that? If not, I could work on a proposal...

scruffian commented:
Does it work to set icon color and size in the theme.json?

RavanH commented:
@scruffian, I tried that but do not think it is possible with this block... IS there any documentation on Social Icons specifically?

scruffian commented:
I think we'll need something like [this](https://github.com/WordPress/gutenberg/issues/35114) for that to work.

scruffian commented:
For now I think we'll have have to make this a setting for the theme...

pbking committed:
Removed requirement to have social parent container be primary menu (or even have content)

pbking commented:
> Now that we have __unstableSocialLinks, why do we need to check for the other attribute? I think having __unstableSocialLinks is specific enough for us to know that we want to render the social links

Pushed a small change with regard to @MaggieCabrera 's comment.  No longer requires the parent container to be parent menu; __unstableSocialLinks on a navigation block is the only requirement.  Holds true for having content; when there is no content the social links are provided (rather than appended) eliminating the need for the empty() check (and allowing the 'social links' menu target to be exclusively targeted as such:

```
<!-- wp:navigation {"__unstableSocialLinks":"social"} /-->
```

I think the existing color management (leveraging `primary` color for social links) is acceptable and appropriate for the time being.

Otherwise (IMO) this change is good-to-go once the appropriate changes have been made to all of the child themes to both add (and style) the social links.

scruffian committed:
Fix layout for social nav

scruffian commented:
I have now added social nav to all Blockbase themes, and updated the CSS to accommodate the new navigation. I have updated the screenshots in the PR description


This should also fix #4590

scruffian committed:
Update the structure of the social links when appended to match the way that the navigation block works

pbking reviewed (approved):
Boom!
All the themes look great.  Seems like a great ponyfill to get things into place.  
Diggit
Shippit
